source/extensions/http/injected_credentials/oauth2/token_provider.cc

```diff
@@ -0,0 +1,128 @@
+#include "source/extensions/http/injected_credentials/oauth2/token_provider.h"
+#include "token_provider.h"
+
+#include <chrono>
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace InjectedCredentials {
+namespace OAuth2 {
+
+namespace {
+
+constexpr absl::string_view DEFAULT_AUTH_SCOPE = "";
+// Transforms the proto list of 'auth_scopes' into a vector of std::string, also
+// handling the default value logic.
+std::string oauthScopesList(const Protobuf::RepeatedPtrField<std::string>& auth_scopes_protos) {
+  std::vector<std::string> scopes;
+
+  // If 'auth_scopes' is empty it must return a list with the default value.
+  if (auth_scopes_protos.empty()) {
+    scopes.emplace_back(DEFAULT_AUTH_SCOPE);
+  } else {
+    scopes.reserve(auth_scopes_protos.size());
+
+    for (const auto& scope : auth_scopes_protos) {
+      scopes.emplace_back(scope);
+    }
+  }
+  return absl::StrJoin(scopes, " ");
+  ;
+}
+
+} // namespace
+
+// TokenProvider Constructor
+TokenProvider::TokenProvider(Common::SecretReaderConstSharedPtr secret_reader,
+                             ThreadLocal::SlotAllocator& tls, Upstream::ClusterManager& cm,
+                             const OAuth2& proto_config, Event::Dispatcher& dispatcher,
+                             const std::string& stats_prefix, Stats::Scope& scope)
+    : secret_reader_(secret_reader), tls_(tls.allocateSlot()),
+      client_id_(proto_config.client_credentials().client_id()),
+      oauth_scopes_(oauthScopesList(proto_config.scopes())), dispatcher_(&dispatcher),
+      stats_(generateStats(stats_prefix + "oauth2.", scope)),
+      retry_interval_(
+          proto_config.token_fetch_retry_interval().seconds() > 0
+              ? std::chrono::seconds(proto_config.token_fetch_retry_interval().seconds())
+              : std::chrono::seconds(2)) {
+  ThreadLocalOauth2ClientCredentialsTokenSharedPtr empty(
+      new ThreadLocalOauth2ClientCredentialsToken(""));
+  tls_->set(
+      [empty](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr { return empty; });
```

Comment:
Thanks for your quick response. I think you can create a timer here directly when you construct the provider. Then all if(timer_) {...} are unnecessary.